Add support for reading YAML from FILE *

sgallagher opened the pull request:
Fixes: https://github.com/fedora-modularity/libmodulemd/issues/50

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

ignatenkobrain commented on modulemd/modulemd-defaults.c:
```diff
@@ -524,6 +524,60 @@ modulemd_defaults_new_from_string (const gchar *yaml_string, GError **error)
+      return NULL;
+    }
+
+  for (gsize i = 0; i < data->len; i++)
```
shouldn't it be `guint`? https://developer.gnome.org/glib/stable/glib-Pointer-Arrays.html#GPtrArray

sgallagher replied:
It's irrelevant, really. gsize is essentially size_t which is pretty much designed for this purpose. It's also signed and always going to be at least as large as guint.

ignatenkobrain commented on modulemd/modulemd-defaults.c:
```diff
@@ -524,6 +524,60 @@ modulemd_defaults_new_from_string (const gchar *yaml_string, GError **error)
+      object = g_ptr_array_index (data, i);
+      if (MODULEMD_IS_DEFAULTS (object))
+        {
+          defaults = MODULEMD_DEFAULTS (g_object_ref (object));
```
I don't think you need `MODULEMD_DEFAULTS` cast here... but not sure.

sgallagher replied:
I do, the compiler warns about it.

ignatenkobrain commented on modulemd/modulemd-defaults.c:
```diff
@@ -524,6 +524,60 @@ modulemd_defaults_new_from_string (const gchar *yaml_string, GError **error)
+modulemd_defaults_new_from_stream (FILE *stream, GError **error)
+{
+  GObject *object = NULL;
+  GPtrArray *data = NULL;
```
`g_autoptr()`?

sgallagher replied:
Ah, I basically just copied this from the `modulemd_defaults_new_from_file()` routine. But you're right; I've been using `g_autoptr()` for all newly-written code; best not to break with that here. I will fix it up.

ignatenkobrain commented on modulemd/modulemd-module.c:
```diff
@@ -3625,6 +3625,53 @@ modulemd_module_new_all_from_string_ext (const gchar *yaml_string,
+modulemd_module_new_from_stream (FILE *stream)
+{
+  GError *error = NULL;
+  ModulemdModule *module = NULL;
```
`g_autoptr`?

sgallagher replied:
Same as above.

sgallagher commented:
I've updated this PR addressing the issues @ignatenkobrain saw, as well as the following:
* Fixed a couple error messages that were copied from the `_file()` versions to say "stream"
* Added a proper error return, even for functions whose file or string versions don't have it. That was a poor design that we're stuck with until 2.0, but there's no reason to keep it for new functions.
